crates/burn-core/src/nn/functional/linear.rs

```diff
@@ -0,0 +1,114 @@
+use crate::tensor::Tensor;
+use crate::tensor::backend::Backend;
+
+/// Applies a linear transformation to the input tensor using the given weight and bias.
+///
+/// ```math
+/// y = x @ weight + [bias]
+/// ```
+///
+/// # Arguments:
+///
+/// - `input` is the input tensor, ``[..., d_input]``.
+/// - `weight` is the weight tensor, ``[d_input, d_output]``.
+/// - `b` is the bias tensor (optional), ``[d_output]``.
+///
+/// # Returns:
+///
+/// The transformed tensor, ``[..., d_output]``.
+///
+/// # Compatibility
+///
+/// This function differs from PyTorch's ``torch.nn.functional.linear`` in that it does not
+/// transpose the weight matrix. In PyTorch, the weight matrix is transposed before
+/// multiplication:
+///
+/// ```math
+/// y = x @ weight^T + [bias]
+/// ```
+pub fn linear<B: Backend, const D: usize>(
+    input: Tensor<B, D>,
+    weight: Tensor<B, 2>,
+    bias: Option<Tensor<B, 1>>,
+) -> Tensor<B, D> {
+    if D == 1 {
+        // Insert and remove an extra batch dimension for the batch matmul to work.
+        return linear::<B, 2>(input.unsqueeze(), weight, bias).flatten(0, 1);
+    }
+
+    let weight = weight.unsqueeze();
+    let output = input.matmul(weight);
+    match bias {
+        Some(bias) => output + bias.unsqueeze(),
+        None => output,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::TestBackend;
+
+    #[test]
+    fn test_linear_1d() {
+        let weight =
+            Tensor::<TestBackend, 2>::from_data([[1.0, 2.0], [3.0, 4.0]], &Default::default());
+
+        let x = Tensor::<TestBackend, 1>::from_data([1.0, 2.0], &Default::default());
+
+        linear(x.clone(), weight.clone(), None)
+            .into_data()
+            .assert_eq(
+                &Tensor::<TestBackend, 1>::from_data([7.0, 10.0], &Default::default()).into_data(),
```

Comment:
You can simplify this as:

```rust
&TensorData::from([7.0, 10.0]),
```
More on this:
https://burn.dev/docs/burn/prelude/struct.TensorData.html#method.from-3